src/cookbook/gradual-composition-api-migration.md

```diff
@@ -0,0 +1,245 @@
+
+## In Place of Vuex
+
+It is possible to use the Composition API in place of Vuex, and save yourself a dependency. That said, it's not exactly necessary, and like many approaches, there are tradeoffs.
+
+If you're using Vuex, it's very clear exactly what centralized state is being used across the application. In addition, you gain the benefit of using a standardized pattern that comes with officially supported tooling in the DevTools which can make a big difference when it comes to debugging code.
+
+The Composition API is very flexible, but you may lose that implicit declaration in communication to fellow maintainers. Our suggestion is that if you do use it as a centralized state management store, that you place it in a `stateManagement` folder, or something similarly named, so that responsibilities are clear.
+
+## Components
+
+It is possible at this point to start to move your components to the Composition API, however, for larger code bases, you may not see immediate need or good reason to, as the Options API works nicely.
+
+Our suggestion is to use this opportunity to think through the architecture and refactor any larger components that have too much responsibility on their own.
+
+For instance, if you have a component that has the following methods:
+
+- one that gathers input from a form
+- one that validates that input
+- one that uses the input to call out to an API
+- one that does something with that API response
+
+You may find that you can potentially create a reusable component for the API call, as on a large application there's a fairly high chance you will need to make another.
+
+For the purposes of demonstration, here's how that one method refactor might look.
+
+### Before, with a Tightly Coupled Method
+
+```js
+methods: {
+  async getGitHubProjects() {
+    try {
+      this.isLoading = true
+      let projects = await fetch(
+        `https://api.github.com/users/sdras/repos?page=1&per_page=100`
+      )
+      this.gitHubProjects = await projects.json()
+    } catch (error) {
+      this.error = error
+      console.error(error);
+    } finally {
+      this.isLoading = false;
+    }
+  }
+}
+```
+
+### After, with the Composition API
+
+```js
+import { reactive, toRefs } from '@vue/composition-api'
+
+export function useFetchAPI(api, options) {
+  const state = reactive({
+    response: null,
+    isLoading: false,
+    hasErrors: null
+  })
+
+  const apiCall = async () => {
+    state.isLoading = true
+    try {
+      let res = await fetch(api, options)
+      state.response = await res.json()
+    } catch (error) {
+      state.hasErrors = error
+    } finally {
+      state.isLoading = false
+    }
+  }
+
+  return {
+    ...toRefs(state),
+    apiCall
+  }
+}
+```
+
+And now we can refactor our earlier component like this:
+
+```js
+import { useFetchAPI } from '@/features/useFetchAPI.js'
+
+export default {
+  data() {
+    return {
+      gitHubData: [],
```

Comment:
With all the changes made, do we still need these reactive properties in data? I mean we could alias and use those we were exposing from `useFetchAPI`.

I understand this kinda makes the method meaningless and kills the point of the example. Maybe we could add one more 'side' data property and side effect in the method changing it? something like `isModalOpen` and change it to false after the API call is resolved?

Reply:
I agree that we can replace the data proprties with the return values of the composition function.

Since that makes the method redundant, maybe the example could instead have a `watch: { errors: ... }` that emits an event when there has been an error?

That would suffiently demonstrate how to access setup return values in Options API

Reply:
good points, I'll refactor. thanks!